### PR TITLE
feat: ability to customize learner builder in vw.h

### PR DESCRIPTION
**Pull request by lalo**
This PR adds the following experimental init fn to vw.h:

```
// Experimental (VW::setup_base_i):
vw* initialize_with_builder(std::string s, io_buf* model = nullptr, bool skipModelLoad = false,
    trace_message_t trace_listener = nullptr, void* trace_context = nullptr,
    std::unique_ptr<VW::setup_base_i> = nullptr);
```

This allows any user of VW to alter the reduction stack or augment it as they find fit. See unit tests for examples of flexibility. They are able to inherit from our concrete impls of setup_base_i or provide their own.

This could also be leveraged by https://github.com/VowpalWabbit/vowpal_wabbit/pull/2971

### CODE REVIEW

File: vowpalwabbit/vw.h [hunk 1]
```diff
@@ -44,13 +44,16 @@ namespace VW
     (2) The code is not yet reentrant.
    */
 vw* initialize(std::unique_ptr<config::options_i, options_deleter_type> options, io_buf* model = nullptr,
-    bool skipModelLoad = false, trace_message_t trace_listener = nullptr, void* trace_context = nullptr);
+    bool skipModelLoad = false, trace_message_t trace_listener = nullptr, void* trace_context = nullptr,
```
Comment (jackgerrits) [inline vowpalwabbit/vw.h:47]: I do not think this should be added to the standard `initialize` calls if it is for testing. Can this be an internal only `initialize` implementation?

Comment (lalo) [reply vowpalwabbit/vw.h:47]: I was planning on refactoring python reduction work to re-use this work, and it would use this same vw.h API. Thoughts?

Comment (lalo) [reply vowpalwabbit/vw.h:47]: it could also be a test and binding only API (controlled via cmake)

Comment (jackgerrits) [reply vowpalwabbit/vw.h:47]: I see, I still think keeping it separate to these entrypoints may make things a little easier. Can it be called `initialize_custom` or something?

File: vowpalwabbit/parse_args.cc [hunk 2]
```diff
@@ -1643,16 +1662,32 @@ vw* initialize(std::string s, io_buf* model, bool skipModelLoad, trace_message_t
   return ret;
 }
 
-vw* initialize_escaped(
-    std::string const& s, io_buf* model, bool skipModelLoad, trace_message_t trace_listener, void* trace_context)
+vw* initialize_with_builder(int argc, char* argv[], io_buf* model, bool skipModelLoad, trace_message_t trace_listener,
+    void* trace_context, std::unique_ptr<VW::setup_base_i> learner_builder)
+{
+  std::unique_ptr<options_i, options_deleter_type> options(
+      new config::options_boost_po(argc, argv), [](VW::config::options_i* ptr) { delete ptr; });
```
Comment (olgavrou) [inline vowpalwabbit/parse_args.cc:1669]: can we use `VW::make_unique` here instead of `new`?

File: vowpalwabbit/parse_args.cc [hunk 3]
```diff
@@ -1408,9 +1408,20 @@ void parse_modules(
...
+// note: Although we have the option to override setup_base_i,
+// the most common scenario is to use the default_reduction_stack_setup.
+// Expect learner_builder to be nullptr most/all of the cases.
+void instantiate_learner(vw& all, std::unique_ptr<VW::setup_base_i> learner_builder)
+{
+  if (!learner_builder)
+  { learner_builder = VW::make_unique<VW::default_reduction_stack_setup>(all, *all.options.get()); }
+  else
+  {
+    learner_builder->delayed_attach(all, *all.options.get());
```
Comment (olgavrou) [inline vowpalwabbit/parse_args.cc:1422]: nit: rename suggestion `delayed_state_attach`

File: vowpalwabbit/vw.h (hunk 1 above)
Comment (lalo) [reply vowpalwabbit/vw.h]: done

File: vowpalwabbit/parse_args.cc [hunk 4]
```diff
@@ -1624,10 +1637,17 @@ vw* initialize(std::unique_ptr<options_i, options_deleter_type> options, io_buf*
   }
 }
 
-vw* initialize(std::string s, io_buf* model, bool skip_model_load, trace_message_t trace_listener, void* trace_context)
+vw* initialize(std::unique_ptr<options_i, options_deleter_type> options, io_buf* model, bool skip_model_load,
```
Comment (lalo) [inline vowpalwabbit/parse_args.cc:1640]: most of these fns have been shuffled around, see vw.h for a diff on what actually changed from the public side.

**Review by jackgerrits — APPROVED** [review]

File: vowpalwabbit/parse_args.cc (hunk 2 above)
Comment (lalo) [reply vowpalwabbit/parse_args.cc:1671]: This is moved code, so we can address in separate PR. Although Jack confirmed that VW::make_unique right now doesn't have support for custom delete.

File: vowpalwabbit/parse_args.cc (hunk 3 above)
Comment (lalo) [reply vowpalwabbit/parse_args.cc]: done

File: vowpalwabbit/reduction_stack.h [hunk 5]
```diff
@@ -23,8 +28,10 @@ struct default_reduction_stack_setup : public setup_base_i
   std::string get_setupfn_name(reduction_setup_fn setup) override;
 
 private:
-  std::vector<std::tuple<std::string, reduction_setup_fn>> reduction_stack;
   VW::config::options_i* options_impl = nullptr;
   vw* all_ptr = nullptr;
+
+protected:
```
[the remark this answers is not in the document]
Comment (lalo) [reply vowpalwabbit/reduction_stack.h:34]: no, this allows anybody to inherit this class and alter the stack without having to figure out where the setup functions of the reductions live. You get access to it on runtime, and before actually running setup. The tests are just demonstrating what's now possible with this new api.

**Review by olgavrou — APPROVED** [review]

**Pull request merged**